packages/@headlessui-vue/src/components/dialog/dialog.ts

```diff
@@ -191,13 +191,28 @@ export let Dialog = defineComponent({
     provide(DialogContext, api)
 
     // Handle outside click
-    useOutsideClick(internalDialogRef, (_event, target) => {
-      if (dialogState.value !== DialogStates.Open) return
-      if (hasNestedDialogs.value) return
+    useOutsideClick(
+      () => {
+        // Third party roots
+        let rootContainers = Array.from(
+          ownerDocument.value?.querySelectorAll('body > *') ?? []
```

Comment:
I don't completely understand what's happening here but I'm curious whether this fix works with elements that are not direct children of the `body` tag? What about clicks on an element nested inside a few others?

Reply:
Alright so the idea with this code is to allow for 3rd party plugins.

We don't allow to interact with elements behind the Dialog component for accessibility reasons and what not. We still don't allow that, but there are a few exceptions that technically violate some of the rules but you can think of it as progressive enhancement.

We already allowed our `Portal` component to be used inside the Dialog (and nested Dialogs). The HTML looks something like this:

```html
<body>
  <div id="app"><!-- Your main application --></div>
  <div id="headlessui-portal-root">
    <div><!-- The main Dialog --></div>
    <div><!-- ... other Portal components --></div>
  </div>
</body>
```

What you will notice is that the other portal components live outside the main Dialog, so technically we should close the Dialog if you click on any of those items because they are "outside" the Dialog component.

The reason this is already possible with our provided `Portal` is because we control that component and we can have a reference to the elements it renders in the DOM.

The issue this PR fixes is with 3rd party plugins. Often 3rd party plugins will render the popup elements in a portal as well. The problem is that we don't know when this happens and we also can't get a DOM reference easily to those elements. We also can't ask every library on planet earth to expose some of the information we need.

So this fix is definitely not perfect, but I think it will solve a lot of the issues people experience today.

How does it work?

Let's imagine you have this structure again:

```html
<body>
  <div id="app">
    <div>
      <!--
        This is an example button in the App that is **not** inside the Dialog.
        Trying to click this button will close the Dialog because this is
        "outside" of the Dialog which is not allowed.
      -->
      <button>in app</button>
    </div>
  </div>
  <div id="headlessui-portal-root">
    <div>
      <!-- This is a button in the Dialog, interacting with this is allowed. -->
      <button>in dialog</button>
    </div>
  </div>
  <div id="third-party-library-portal">
    <!--
      Interacting with this button is technically not allowed since it lives
      outside of the Dialog. However we make an exception that you _can_
      interact with this one because it lives in another "parent" than the main
      application. This means that we are currently making an assumption that
      interacteable elements that live in a parent outside of the main app are allowed.

      This trade-off is necessary since we don't know when 3rd party libraries
      will render certain elements in the DOM and we don't get a stable
      reference to those elements.
    -->
    <button>Deeply nested button inside the 3rd party library</button>
  </div>
</body>
```

How it works is that we collect all the direct "root" containers, that's what the `body > *` is for. In this case we will get 3 elements:

- <div id="app">
- <div id="headlessui-portal-root">
- <div id="third-party-library-portal">

- Then we filter out the roots we know about, in this case we don't allow interactions in the `<div id="app">` beacuse that would violate our initial rule of "don't interact with elements outside of the Dialog".
- Next, we still run the `outside click` logic, this function gives us a `target` which is the target element that we are clicking.
- Last but not least we check if one of those root containers `contains` the `target`. The [`contains` function](https://developer.mozilla.org/en-US/docs/Web/API/Node/contains) checks whether one element is contained within another (also does check deeply nested).
- This means that:
  - If the `target` is contained inside the `<div id="app">` then we should run the outside click callback which in this case closes the Dialog
  - But if the `target` is contained inside the `<div id="third-party-library-portal">` then that's fine because we assume that this is the 3rd party library.

Does that answer your question?

Reply:
Wow yeah that does answer my question, thanks! I really appreciate the detailed explanation. At one point I had thought it would be easier to just listen for clicks directly on the backdrop, but now I'm realizing there doesn't even necessarily have to be a backdrop... not to mention all the other possibilities. This is way more complicated than I thought and your solution looks great. Thanks for breaking it down 👍🏻